src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,324 @@
+  child.stderr.destroy();
+  child.unref();
+  child.disconnect();
+
+  const flavor = instanceInfo.flavor;
+  const { host, port } = instanceInfo.server;
+  const cmd =
+    process.platform === "win32"
+      ? path.basename(process.execPath)
+      : [process.execPath, ...process.execArgv, module, ...childArgs].join(" ");
+  const pid = child.pid;
+  const startTime = Date.now();
+
+  const instance: DetachedInstance = {
+    startTime,
+    pid,
+    name: createInstanceName(),
+    host,
+    port,
+    flavor,
+    cmd,
+    version
+  };
+
+  while (true) {
+    const instanceFilename = getInstanceFilePath(instance.name);
+    try {
+      await writeFile(instanceFilename, JSON.stringify(instance), {
+        // wx means "Open file for writing, but fail if the path exists". see
+        // https://nodejs.org/api/fs.html#file-system-flags
+        flag: "wx",
+        encoding: FILE_ENCODING
+      });
+      break;
+    } catch (err) {
+      switch ((err as NodeJS.ErrnoException).code) {
+        case "EEXIST":
+          // an instance already exists with this name
+          instance.name = createInstanceName();
+          break;
+        case "ENOENT":
+          // we don't check whether the folder exists before writing, as that's
+          // a very uncommon case. Catching the exception and subsequently
+          // creating the directory is faster in the majority of cases.
+          await mkdir(dataPath, { recursive: true });
+          break;
+        default:
+          throw err;
+      }
+    }
+  }
+
+  return instance;
+}
+
+/**
+ * Fetch all instance of Ganache running in detached mode. Cleans up any
+ * instance files for processes that are no longer running.
+ * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
+ */
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  let dirEntries: Dirent[];
+  let processes: ProcessDescriptor[];
+  let someInstancesFailed = false;
+
+  try {
+    [dirEntries, processes] = await Promise.all([
+      readdir(dataPath, { withFileTypes: true }),
+      psList()
+    ]);
+  } catch (err) {
+    if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
+      throw err;
+    }
+    // instances directory does not (yet) exist, so there cannot be any instances
+    return [];
+  }
+  const instances: DetachedInstance[] = [];
+
+  const loadingInstancesInfos = dirEntries.map(async dirEntry => {
+    const filename = dirEntry.name;
+    const { name: instanceName, ext } = path.parse(filename);
+
+    let failureReason: string;
+
+    if (ext !== ".json") {
+      failureReason = `"${filename}" does not have a .json extension`;
+    } else {
+      let instance: DetachedInstance;
+      try {
+        // getDetachedInstanceByName() throws if the instance file is not found or
+        // cannot be parsed
+        instance = await getDetachedInstanceByName(instanceName);
+      } catch (err: any) {
+        failureReason = err.message;
```

Comment:
This is likely going to be a JSON parsing error - which is fairly informative, or a "file not found" or "access denied" sort of a thing.